### ISSUE TITLE
photon shooting?

### ISSUE THREAD

**Issue by beckermr** [GalSim-developers/JAX-GalSim#50]
How do we feel about photon shooting? 

some initial thoughts:
 
- GalSim has its own RNGs and jax requires you to use their RNG system
- from a technical perspective, photon shooting is not so bad to get right
- gradients through images made with photon shooting will be very odd / will not work since the binning into the pixels is discrete
- however from the point of view of making images, being able to photon shoot on the GPUs would potentially be very efficient

BTW, I am very impressed by the work here everyone!

**Comment by EiffL** [comment]
I have nothing against photon shooting :-) 

I'm not too worried about the stochastic aspect of it. We have to replace GalSim's internal RNG system by a jax-based one anyway (which I think we can do while keeping the same API). 

But yeah to make it differentiable we might want to use some relaxation tricks (essentially making the pixel assignment continuous and not discrete, so blurring a bit the pixel edges).

Also, I have NOT looked at the photon shooting implementation in GalSim, I'm not sure how it works in practice. Like do you draw from a base profile, then apply jacobian transformations for each photons, and somehow also apply the deviations caused by the PSF? 
Or do you get the surface brightness profile after psf convolution, and just draw photons from that profile?

**Comment by beckermr** [comment]
The psf is added as a random perturbation to the photon locations. I agree the relaxation tricks would be good. We should add those as an option.

**Comment by EiffL** [comment]
Interesting... Then yeah sounds very doable :-)

**Comment by EiffL** [comment]
Ah... one more serious complication. We need to know the number of photons we are going to draw at JIt time... And because that will depend on the magnitude of the object (which we certainly want to keep as a dynamically traced parameter) we can't know it. 
The simple (but very ugly) solution is to have a fixed array of size N_max photons, and then give a weight to each photon (0 or 1) . this solution maintains fixed dimensionality, but of course is problematic for other reasons....

**Comment by beckermr** [comment]
Yup good point. I'm skeptical that making photon shooting differentiable is going to be worth the time and trouble in the end. I'd be tempted to just give up on the first pass and use while loops which can have variable sized lengths.

**Comment by EiffL** [comment]
Everything is differentiable if you are brave enough :-D 
We might be able to do this in a clever way, like drawing by batches with a `jax.lax.scan` the batches can have fixed sizes, and the last batch can be partially filled, with some photons having zero weight, so this way you reduce the memory overhead.

**Comment by beckermr** [comment]
@rmjarvis Can explain in detail galsim has a flux per photon? It seems logical, but I feel like I am missing some details in how this is used internally.

**Comment by rmjarvis** [comment]
Normally it's 1 e- per photon (which is 1 ADU per photon with the default gain=1).  The primary exception is InterpolatedImage, which can have negative flux parts.  This implies that some of the photons need to have negative flux to implement this properly.  To get the right Poisson statistics, the positive (and negative) flux photons can't have flux=1.  There is a detailed discussion of this in section 6.3.6 of [Rowe et all](https://ui.adsabs.harvard.edu/abs/2015A%26C....10..121R/abstract).  Also in the code comment [here](https://github.com/GalSim-developers/GalSim/blob/releases/2.4/galsim/gsobject.py#L2077).

**Comment by EiffL** [comment]
Sorry for this very noob question, but can I ask what are the advantages of photon shooting? 
FFTs are pretty cheap on GPU, do they have particular limitations that photon shooting gets around?

**Comment by rmjarvis** [comment]
It depends on how bright the object is.  Shooting ~10 photons is MUCH faster than doing FFTs.  In realistic sims, we render lots of objects with flux<100.

**Comment by rmjarvis** [comment]
Also, if you care about wavelength-dependent effects, photon shooting is significantly more efficient at almost any flux level.

**Comment by ismael-mendoza** [comment]
Closed by #82
